### PR TITLE
Use ORDER BY to inform index selection.

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

File: sql/structured.go [hunk 1]
```diff
@@ -138,6 +138,18 @@ func (desc *IndexDescriptor) containsColumnID(colID ColumnID) bool {
 	return false
 }
 
+// fullColumnIDs returns the index column IDs including any implicit column IDs
+// for non-unique indexes.
+func (desc *IndexDescriptor) fullColumnIDs(tableDesc *TableDescriptor) []ColumnID {
```
Comment (tamird) [inline sql/structured.go]: nit: slight preference for making this a method on the table descriptor. also this method should assert that this index actually belongs to the table

File: sql/select.go [hunk 2]
```diff
@@ -121,41 +120,52 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
 // the query (we currently only support covering indexes) and selects the
 // "best" index from that set. The cost model based on keys per row, key size
 // and number of index elements used.
-func (p *planner) selectIndex(s *scanNode) (planNode, error) {
-	if s.desc == nil || s.filter == nil {
-		// No table or where-clause.
-		s.initOrdering()
-		return s, nil
+func (p *planner) selectIndex(scan *scanNode, ordering []int) (planNode, error) {
```
Comment (tamird) [inline sql/select.go]: can you do this rename in a separate commit? quite noisy currently

File: sql/structured.go (hunk 1 above)
Comment (petermattis) [inline sql/structured.go]: I went back and forth on this. Any reason for that preference? What would you name this method on `TableDescriptor`? `fullIndexColumnIDs`?

File: sql/explain.go [hunk 3]
```diff
@@ -48,14 +48,25 @@ func (p *planner) Explain(n *parser.Explain) (planNode, error) {
 	if err != nil {
 		return nil, err
 	}
+	plan, err = markExplain(plan, mode)
+	if err != nil {
+		return nil, fmt.Errorf("%v: %s", err, n)
```
Comment (tamird) [inline sql/explain.go]: are those verbs reversed?

File: sql/structured.go (hunk 1 above)
Comment (tamird) [inline sql/structured.go]: Yeah, that's a reasonable name. I'm actually fine with this as-is, but the assertion seems important.

File: sql/explain.go (hunk 3 above)
Comment (petermattis) [inline sql/explain.go]: No. I usually use `%v` for err because it will print `<nil>` if `err == nil`. `%s` is the correct verb for printing a parse node.

**Comment by tamird** [comment]
Metacomment here: there's a fair amount of preparation here in addition to the actual addition that is the meat of the PR. In the future it would be super helpful to separate those changes for ease of review.

File: sql/select.go (hunk 2 above)
Comment (petermattis) [inline sql/select.go]: Hrmm, side-by-side diff makes the renaming portions more obvious.

File: sql/explain.go (hunk 3 above)
Comment (tamird) [inline sql/explain.go]: alright

File: sql/select.go (hunk 2 above)
Comment (petermattis) [inline sql/select.go]: I'll just revert the renaming in this method.

File: sql/structured.go (hunk 1 above)
Comment (petermattis) [inline sql/structured.go]: I have a mild preference to leave the method here. I'll add the assertion.

File: sql/select.go [hunk 4]
```diff
@@ -121,19 +120,14 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
 // the query (we currently only support covering indexes) and selects the
 // "best" index from that set. The cost model based on keys per row, key size
 // and number of index elements used.
-func (p *planner) selectIndex(s *scanNode) (planNode, error) {
-	if s.desc == nil || s.filter == nil {
-		// No table or where-clause.
+func (p *planner) selectIndex(s *scanNode, ordering []int) (planNode, error) {
+	if s.desc == nil || (s.filter == nil && ordering == nil) {
```
Comment (tamird) [inline sql/select.go]: i might not be seeing it, but it seems that `ordering` can never be nil since it's only ever passed in from `sort.Ordering()` above `sortNode`s are never constructed with a nil slice.

**Comment by petermattis** [comment]
Not sure if I would qualify the minor reorgs as a fair amount of preparation work. And those reorgs happened after the meat of the PR. I need better git-fu to be able to separate them out into separate commits.

File: sql/select.go (hunk 4 above)
Comment (petermattis) [inline sql/select.go]: If `sortNode` is `nil`, `sortNode.Ordering()` returns nil. Perhaps this was too clever.

Comment (tamird) [inline sql/select.go]: yeah, it's a little clever; i'd add a comment at least

File: sql/testdata/order_by [hunk 5]
```diff
@@ -118,3 +117,36 @@ SELECT * FROM t ORDER BY foo
...
+  a INT,
+  b INT,
+  c INT,
+  PRIMARY KEY (a, b, c),
+  CONSTRAINT bc INDEX (b, c),
+  CONSTRAINT ba INDEX (b, a)
+)
+
+statement ok
+INSERT INTO abc VALUES (1, 2, 3)
```
Comment (tamird) [inline sql/testdata/order_by]: it may be illustrative to add another row here to demonstrate that we scan forward (and sort in-memory) even when reverse order is requested

**Comment by tamird** [comment]
LGTM

File: sql/testdata/order_by (hunk 5 above)
Comment (petermattis) [inline sql/testdata/order_by]: Ok. Another row added and two more test cases demonstrating that we currently always scan forward. And added a TODO that about the test breakage that will occur when reverse scan is implemented.
